contents/javascript/staticTypingWithTypescript.md

```diff
@@ -0,0 +1,157 @@
+
+- Static type checking can **make developers more productive**. When paired with the appropriate IDE support, it enables features such as auto-completion and offers instant feedback by incrementally rechecking your code as you type.
+
+- Static type checking **improves collaboration on a large codebase**. Explicitly declared types allow developers to refactor with greater confidence, knowing that API boundaries are enforced by the type checker. If you change a method signature or the shape of an object, you can quickly discover the other parts of the codebase that also need to be changed, as they would be surfaced as errors by the type checker.
+
+
+### What's the difference between Flow and TypeScript?
+
+**[Flow](https://flow.org/)** is a static type checker for JavaScript. It is an open-source tool developed by Facebook.
+
+**[TypeScript](https://www.typescriptlang.org/)** is a statically typed superset of JavaScript that compiles to plain JavaScript. It is an open-source programming language developed by Microsoft.
+
+While Flow and TypeScript integrate into your development workflow in slightly different ways, they have the same goals and share many similarities in terms of syntax. As the industry seems to be shifting towards TypeScript as the top choice <sup>[source](https://dev.to/nickytonline/is-2019-the-year-of-typescript-18p2)</sup>, in this article I will introduce how TypeScript works and how you can get started with TypeScript.
+
+
+## TypeScript Basics
+
+### Compilation to JavaScript
+
+As a superset of JavaScript, TypeScript is based on the same programming syntax as JavaScript, and valid JavaScript code is also valid TypeScript code. Besides static type checking, TypeScript offers support for the latest JavaScript features, including those from ECMAScript 2015 (ES6) and beyond.
+
+While TypeScript can be configured to target different flavors of JavaScript, it is transpiled down to ES3 by default to maintain compatibility with all JavaScript engines. Because of this, you can use features like classes, generators, and async functions without worrying about whether your code is supported by a particular web browser or version of Node.js.
+
+
+### Type Annotations and Type Inference
+
+A key feature of TypeScript is the ability to add **type annotations** to your JavaScript variables and functions. This is useful for recording the intended contract of a variable or function.
+
+Here's an example of declaring a variable with the `string` type and assigning it a `string` value:
+```typescript
+let myString: string = "abc";
+```
+
+If we try to assign it a `number` value later on, the TypeScript compiler will raise an error:
+```typescript
+myString = 123; // error: 123 is not assignable to type 'string'
```

Comment:
Haha I definitely considered this. There a few reasons why I decided against it eventually.

1) Readers won't be able to copy-paste the code snippets if they wanna play with the examples themselves - I'm going to introduce https://www.typescriptlang.org/play/ in the last section.

2) If I hover over the error, it will block the rest of the code. Of course, I can work around this by adding more line breaks above the errored line, but it'll be uglier. I also want to simplify the error messages, so using comments would be better than letting readers see the actual IDE errors. This is especially true in the case of objects.

<img width="671" alt="Screenshot 2019-04-20 at 1 50 31 PM" src="https://user-images.githubusercontent.com/9073504/56453314-516d0980-6373-11e9-9d34-c88ed4e68840.png">

3) Screenshots are much harder to update later on if someone else wants make any changes / additions to the article. A new screenshot might have different syntax highlighting, different font, etc. which makes the article look less consistent overall.